Emit all auth token changes to subscribers

ryanjduffy opened the pull request:
Tracks all the open replay.io tabs and dispatches token changes to each

ryanjduffy committed:
Emit all auth token changes to subscribers

jcmorrow reviewed (commented):
This all looks like fine to me although I expect @loganfsmyth to have more helpful context!

loganfsmyth commented on devtools/server/actors/replay/auth.js:
```diff
@@ -125,11 +138,8 @@ function handleAuthChannelMessage(_id, message, target) {
   if (type === "connect") {
-    deferredAccessToken.promise.then(token => {
-      if (authChannel) {
-        authChannel.send({ token }, target);
```
I'm actually a bit surprised here about how this was working previously, because `registerWebChannel` assigns this, except we call `registerWebChannel` twice and the second one is with `localUrl` which means this was probably always using the wrong `authChannel` (with the right `target`).

We should adjust this so that `handleAuthChannelMessage` knows what channel it is listening to, and we need to make sure we always use that channel to reply to the `connect` message?

loganfsmyth commented on devtools/server/actors/replay/auth.js:
```diff
@@ -117,6 +107,29 @@ function tokenExpiration(token) {
+      if (target.browsingContext.currentWindowGlobal) {
+        notifyWebChannelTarget(target);
+      } else {
+        webChannelTargets.delete(key)
```
This will delete the Map entry during iteration meaning I have no idea what this `for` loop will do.
```diff
-for (let [key, target] of webChannelTargets.entries()) {
+for (let [key, target] of Array.from(webChannelTargets.entries())) {
```
please!

ryanjduffy replied:
I did verify that the iterator is okay with deleting entries while iterating so I don't think this is necessary.

loganfsmyth replied:
Hmm, I guess you're technically right, I just find this terrifying. Any mutation of something during iteration should be avoided in my opinion.

jcmorrow replied:
This is a side note but I feel like an under-rated solution to these kinds of problems is iterating backwards 😛

loganfsmyth commented on devtools/server/actors/replay/auth.js:
```diff
@@ -73,12 +62,13 @@ function setReplayRefreshToken(token) {
+  if (token === getReplayUserToken()) return;
+
+  Services.prefs.setStringPref("devtools.recordreplay.user-token", token);
+  notifyWebChannelTargets();
```
Would we care enough to do this with
```
Services.prefs.addObserver("devtools.recordreplay.user-token", () => {
  notifyWebChannelTargets();
});
```
so it also applies if the user clears the token in `about:config` or something?

ryanjduffy replied:
That seems reasonable to me. :+1:

ryanjduffy committed:
feedback

ryanjduffy committed:
convert iterator to array

ryanjduffy commented:
* Removed the module-scoped `authChannel` in favor of passing the channel to the handler so we ensure we're talking to the right thing
* Added a pref listener to ensure the auth state is synced if the pref is manually cleared
